File: sdks/php/rest/src/BrumeClient.php
<?php

namespace Brume\Rest;

use Brume\Rest\Public_\PublicClient;
use Brume\Rest\Stats\StatsClient;
use Brume\Rest\ApiKeys\ApiKeysClient;
use Brume\Rest\Channels\ChannelsClient;
use Brume\Rest\Longpoll\LongpollClient;
use Brume\Rest\Postgres\PostgresClient;
use Brume\Rest\Projects\ProjectsClient;
use Brume\Rest\Sse\SseClient;
use Brume\Rest\Webhooks\WebhooksClient;
use Psr\Http\Client\ClientInterface;
use Brume\Rest\Core\Client\RawClient;
use Brume\Rest\Exceptions\BrumeException;
use Brume\Rest\Exceptions\BrumeApiException;
use Brume\Rest\Core\Json\JsonApiRequest;
use Brume\Rest\Core\Client\HttpMethod;
use Brume\Rest\Core\Json\JsonDecoder;
use JsonException;
use Psr\Http\Client\ClientExceptionInterface;
use Exception;

class BrumeClient
{
    /**
     * @var PublicClient $public_
     */
    public PublicClient $public_;

    /**
     * @var StatsClient $stats
     */
    public StatsClient $stats;

    /**
     * @var ApiKeysClient $apiKeys
     */
    public ApiKeysClient $apiKeys;

    /**
     * @var ChannelsClient $channels
     */
    public ChannelsClient $channels;

    /**
     * @var LongpollClient $longpoll
     */
    public LongpollClient $longpoll;

    /**
     * @var PostgresClient $postgres
     */
    public PostgresClient $postgres;

    /**
     * @var ProjectsClient $projects
     */
    public ProjectsClient $projects;

    /**
     * @var SseClient $sse
     */
    public SseClient $sse;

    /**
     * @var WebhooksClient $webhooks
     */
    public WebhooksClient $webhooks;

    /**
     * @var array{
     *   baseUrl?: string,
     *   client?: ClientInterface,
     *   maxRetries?: int,
     *   timeout?: float,
     *   headers?: array<string, string>,
     * } $options @phpstan-ignore-next-line Property is used in endpoint methods via HttpEndpointGenerator
     */
    private array $options;

    /**
     * @var RawClient $client
     */
    private RawClient $client;

    /**
     * @param ?string $apiKey The apiKey to use for authentication.
     * @param ?array{
     *   baseUrl?: string,
     *   client?: ClientInterface,
     *   maxRetries?: int,
     *   timeout?: float,
     *   headers?: array<string, string>,
     * } $options
     */
    public function __construct(
        ?string $apiKey = null,
        ?array $options = null,
    ) {
        $apiKey ??= $this->getFromEnvOrThrow('BRUME_API_KEY', 'Please pass in apiKey or set the environment variable BRUME_API_KEY.');
        $defaultHeaders = [
            'Authorization' => "Bearer $apiKey",
            'X-Fern-Language' => 'PHP',
            'X-Fern-SDK-Name' => 'Brume\Rest',
        ];

        $this->options = $options ?? [];

        $this->options['headers'] = array_merge(
            $defaultHeaders,
            $this->options['headers'] ?? [],
        );

        $this->client = new RawClient(
            options: $this->options,
        );

        $this->public_ = new PublicClient($this->client, $this->options);
        $this->stats = new StatsClient($this->client, $this->options);
        $this->apiKeys = new ApiKeysClient($this->client, $this->options);
        $this->channels = new ChannelsClient($this->client, $this->options);
        $this->longpoll = new LongpollClient($this->client, $this->options);
        $this->postgres = new PostgresClient($this->client, $this->options);
        $this->projects = new ProjectsClient($this->client, $this->options);
        $this->sse = new SseClient($this->client, $this->options);
        $this->webhooks = new WebhooksClient($this->client, $this->options);
    }

    /**
     * Returns the full operational picture for a project: counters,
     * latency percentiles, Postgres lag, top channels, and a plan-limit
     * snapshot. Project API key auth (any scope). No secrets, no
     * high-cardinality data, no admin gating.
     *
     * Response body is documented as `additionalProperties: true` for
     * the same reason as `/v1/stats` and `/v1/analytics`.
     *
     * @param ?array{
     *   baseUrl?: string,
     *   maxRetries?: int,
     *   timeout?: float,
     *   headers?: array<string, string>,
     *   queryParameters?: array<string, mixed>,
     *   bodyProperties?: array<string, mixed>,
     * } $options
     * @return ?array<string, mixed>
     * @throws BrumeException
     * @throws BrumeApiException
     */
    public function diagnostics(?array $options = null): ?array
    {
        $options = array_merge($this->options, $options ?? []);
        try {
            $response = $this->client->sendRequest(
                new JsonApiRequest(
                    baseUrl: $options['baseUrl'] ?? $this->client->options['baseUrl'] ?? '',
                    path: "v1/diagnostics",
                    method: HttpMethod::GET,
                ),
                $options,
            );
            $statusCode = $response->getStatusCode();
            if ($statusCode >= 200 && $statusCode < 400) {
                $json = $response->getBody()->getContents();
                if (empty($json)) {
                    return null;
                }
                return JsonDecoder::decodeArray($json, ['string' => 'mixed']); // @phpstan-ignore-line
            }
        } catch (JsonException $e) {
            throw new BrumeException(message: "Failed to deserialize response: {$e->getMessage()}", previous: $e);
        } catch (ClientExceptionInterface $e) {
            throw new BrumeException(message: $e->getMessage(), previous: $e);
        }
        throw new BrumeApiException(
            message: 'API request failed',
            statusCode: $statusCode,
            body: $response->getBody()->getContents(),
        );
    }

    /**
     * @param string $env
     * @param string $message
     * @return string
     */
    private function getFromEnvOrThrow(string $env, string $message): string
    {
        $value = getenv($env);
        return $value ? (string) $value : throw new Exception($message);
    }
}

File: sdks/swift/rest/Sources/BrumeClient.swift
import Foundation

/// Use this class to access the different functions within the SDK. You can instantiate any number of clients with different configuration that will propagate to these functions.
public final class BrumeClient: Sendable {
    public let `public`: PublicClient
    public let stats: StatsClient
    public let apiKeys: ApiKeysClient
    public let channels: ChannelsClient
    public let longpoll: LongpollClient
    public let postgres: PostgresClient
    public let projects: ProjectsClient
    public let sse: SseClient
    public let webhooks: WebhooksClient
    private let httpClient: HTTPClient

    /// Initialize the client with the specified configuration and a static bearer token.
    ///
    /// - Parameter baseURL: The base URL to use for requests from the client. If not provided, the default base URL will be used.
    /// - Parameter apiKey: Bearer token for authentication. If provided, will be sent as "Bearer {token}" in Authorization header.
    /// - Parameter headers: Additional headers to send with each request.
    /// - Parameter timeout: Request timeout in seconds. Defaults to 60 seconds. Ignored if a custom `urlSession` is provided.
    /// - Parameter maxRetries: Maximum number of retries for failed requests. Defaults to 2.
    /// - Parameter urlSession: Custom `URLSession` to use for requests. If not provided, a default session will be created with the specified timeout.
    public convenience init(
        baseURL: String,
        apiKey: String,
        headers: [String: String]? = nil,
        timeout: Int? = nil,
        maxRetries: Int? = nil,
        urlSession: Networking.URLSession? = nil
    ) {
        self.init(
            baseURL: baseURL,
            headerAuth: nil,
            bearerAuth: .init(token: .staticToken(apiKey)),
            basicAuth: nil,
            headers: headers,
            timeout: timeout,
            maxRetries: maxRetries,
            urlSession: urlSession
        )
    }

    /// Initialize the client with the specified configuration and an async bearer token provider.
    ///
    /// - Parameter baseURL: The base URL to use for requests from the client. If not provided, the default base URL will be used.
    /// - Parameter apiKey: An async function that returns the bearer token for authentication. If provided, will be sent as "Bearer {token}" in Authorization header.
    /// - Parameter headers: Additional headers to send with each request.
    /// - Parameter timeout: Request timeout in seconds. Defaults to 60 seconds. Ignored if a custom `urlSession` is provided.
    /// - Parameter maxRetries: Maximum number of retries for failed requests. Defaults to 2.
    /// - Parameter urlSession: Custom `URLSession` to use for requests. If not provided, a default session will be created with the specified timeout.
    public convenience init(
        baseURL: String,
        apiKey: @escaping ClientConfig.CredentialProvider,
        headers: [String: String]? = nil,
        timeout: Int? = nil,
        maxRetries: Int? = nil,
        urlSession: Networking.URLSession? = nil
    ) {
        self.init(
            baseURL: baseURL,
            headerAuth: nil,
            bearerAuth: .init(token: .provider(apiKey)),
            basicAuth: nil,
            headers: headers,
            timeout: timeout,
            maxRetries: maxRetries,
            urlSession: urlSession
        )
    }

    init(
        baseURL: String,
        headerAuth: ClientConfig.HeaderAuth? = nil,
        bearerAuth: ClientConfig.BearerAuth? = nil,
        basicAuth: ClientConfig.BasicAuth? = nil,
        headers: [String: String]? = nil,
        timeout: Int? = nil,
        maxRetries: Int? = nil,
        urlSession: Networking.URLSession? = nil
    ) {
        let config = ClientConfig(
            baseURL: baseURL,
            headerAuth: headerAuth,
            bearerAuth: bearerAuth,
            basicAuth: basicAuth,
            headers: headers,
            timeout: timeout,
            maxRetries: maxRetries,
            urlSession: urlSession
        )
        self.public = PublicClient(config: config)
        self.stats = StatsClient(config: config)
        self.apiKeys = ApiKeysClient(config: config)
        self.channels = ChannelsClient(config: config)
        self.longpoll = LongpollClient(config: config)
        self.postgres = PostgresClient(config: config)
        self.projects = ProjectsClient(config: config)
        self.sse = SseClient(config: config)
        self.webhooks = WebhooksClient(config: config)
        self.httpClient = HTTPClient(config: config)
    }

    /// Returns the full operational picture for a project: counters,
    /// latency percentiles, Postgres lag, top channels, and a plan-limit
    /// snapshot. Project API key auth (any scope). No secrets, no
    /// high-cardinality data, no admin gating.
    /// 
    /// Response body is documented as `additionalProperties: true` for
    /// the same reason as `/v1/stats` and `/v1/analytics`.
    ///
    /// - Parameter requestOptions: Additional options for configuring the request, such as custom headers or timeout settings.
    public func diagnostics(requestOptions: RequestOptions? = nil) async throws -> [String: JSONValue] {
        return try await httpClient.performRequest(
            method: .get,
            path: "/v1/diagnostics",
            requestOptions: requestOptions,
            responseType: [String: JSONValue].self
        )
    }
}